contracts/Joint.sol

```diff
@@ -22,123 +28,129 @@ interface IERC20Extended {
     function symbol() external view returns (string memory);
 }
 
+interface ProviderStrategy {
+    function vault() external view returns (VaultAPI);
+
+    function strategist() external view returns (address);
+
+    function keeper() external view returns (address);
+
+    function want() external view returns (address);
+}
+
 contract Joint {
     using SafeERC20 for IERC20;
     using Address for address;
     using SafeMath for uint256;
 
+    ProviderStrategy public providerA;
+    ProviderStrategy public providerB;
+
     address public tokenA;
-    address public providerA;
     address public tokenB;
-    address public providerB;
+
     bool public reinvest;
 
-    address public governance;
-    address public pendingGovernance;
-    address public keeper;
-    address public strategist;
     address public WETH;
     address public reward;
     address public router;
+
     uint256 public pid;
-    uint256 public ratio = 500;
-    uint256 public constant MAX_RATIO = 1000;
+
     IMasterchef public masterchef;
 
-    modifier onlyGov {
-        require(msg.sender == governance);
+    IUniswapV2Pair public pair;
+
+    uint256 private investedA;
+    uint256 private investedB;
+
+    modifier onlyGovernance {
+        require(
+            msg.sender == providerA.vault().governance() ||
+                msg.sender == providerB.vault().governance()
+        );
         _;
     }
 
-    modifier onlyGovOrStrategist {
-        require(msg.sender == governance || msg.sender == strategist);
+    modifier onlyAuthorized {
+        require(
+            msg.sender == providerA.vault().governance() ||
+                msg.sender == providerB.vault().governance() ||
+                msg.sender == providerA.strategist() ||
+                msg.sender == providerB.strategist()
+        );
+        _;
+    }
+
+    modifier onlyKeepers {
```

Comment:
not used.

Reply:
Should we add some management modifiers?

Reply:
I don't think it will be necessary. kill it